direct/core/src/main/java/cz/o2/proxima/direct/batch/BatchLogReaders.java

```diff
@@ -115,34 +115,45 @@ private ForwardingBatchLogObserver(BatchLogObserver delegate) {
 
   private static class ThroughputLimitedBatchLogObserver extends ForwardingBatchLogObserver {
 
-    private final int numPartitions;
     private final Collection<Partition> assignedPartitions;
     private final ThroughputLimiter limiter;
     private long watermark = Long.MIN_VALUE;
 
     public ThroughputLimitedBatchLogObserver(
         BatchLogObserver delegate,
-        int numPartitions,
         Collection<Partition> assignedPartitions,
         ThroughputLimiter limiter) {
 
       super(delegate);
-      this.numPartitions = numPartitions;
       this.assignedPartitions = new ArrayList<>(assignedPartitions);
-      this.limiter = Objects.requireNonNull(limiter);
+      this.limiter = SerializableUtils.clone(Objects.requireNonNull(limiter));
```

Comment:
same as in commit log observer

Reply:
yes

Reply:
This is actually correct. The problem is that `LogReader` (which owns the `ThroughputLimiter`) has different life-cycle than the `LogObserver`. You can have multiple observers observe the same reader and therefore each needs its own instance of limiter.